File: library/src/main/java/com/jdroid/java/utils/CopyUtils.java
package com.jdroid.java.utils;

import com.jdroid.java.exception.UnexpectedException;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;

public class CopyUtils {

	@SuppressWarnings("unchecked")
	public static <T extends Serializable> T cloneSerializable(T object) {
		T cloneObject;

		try {
			ByteArrayOutputStream bout = new ByteArrayOutputStream();
			ObjectOutputStream oout = new ObjectOutputStream(bout);
			oout.writeObject(object);
			byte[] bytes = bout.toByteArray();
			oout.close();

			ByteArrayInputStream bin = new ByteArrayInputStream(bytes);
			ObjectInputStream oin = new ObjectInputStream(bin);
			cloneObject = (T)oin.readObject();
			oin.close();
		} catch (IOException | ClassNotFoundException e) {
			throw new UnexpectedException("Failed to clone " + object.getClass().getSimpleName()
					+ " using serialization", e);
		}

		return cloneObject;
	}
}

File: core-legacy/src/main/java/com/dipien/core/utils/CopyUtils.kt
package com.dipien.core.utils

import com.dipien.core.exception.UnexpectedException
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.io.Serializable

object CopyUtils {

    @Suppress("UNCHECKED_CAST")
    fun <T : Serializable> cloneSerializable(serializable: T): T {
        val cloneObject: T
        try {
            val bout = ByteArrayOutputStream()
            val oout = ObjectOutputStream(bout)
            oout.writeObject(serializable)
            val bytes = bout.toByteArray()
            oout.close()

            val bin = ByteArrayInputStream(bytes)
            val oin = ObjectInputStream(bin)
            cloneObject = oin.readObject() as
                T
            oin.close()
        } catch (e: IOException) {
            throw UnexpectedException(
                "Failed to clone " + serializable.javaClass.simpleName + " using serialization", e
            )
        } catch (e: ClassNotFoundException) {
            throw UnexpectedException("Failed to clone " + serializable.javaClass.simpleName + " using serialization", e)
        }
        return cloneObject
    }
}
